new ESM implementation (#26745)
nodejs/node: doc/api/cli.md

Enable FIPS-compliant crypto at startup. (Requires Node.js to be built with
`./configure --openssl-fips`.)

### `--es-module-specifier-resolution=mode`
<!-- YAML
added: REPLACEME
-->

> To be used in conjunction with `--experimental modules`. Sets the resolution

Reviewer: Please add `-`

```diff 
- --experimental modules
+ --experimental-modules
```

algorithm for resolving specifiers. Valid options are `explicit` and `node`.

> The default is `explicit`, which requires you to provide the full path to a

Reviewer 2: Please avoid "you" in the docs

```suggestion
The default is `explicit`, which requires providing the full path to a
```

Track heap object allocations for heap snapshots.

> ### `--type=type`

Reviewer 3: It seems kind of odd that the flag `--type` without any prefixes is dedicated to ESM, is this expected to change when we unflag `--experimental-modules`?

Author: There is a conversation going on right now within the modules team about whether or not we want to change the name of the `type` flag. It was meant to mirror `package.type`, which in itself is based on `<script type="module">`.

We may need to fix the docs because the flag requires and argument

Reviewer 2: Yeah, I have to agree that type is far too generic here. This definitely should be changed.

Reviewer 4: There was some discussion of this before, but there was not consensus to change it. (Personally I've love for it to be renamed.)

I believe `--entry-type` was the specific name proposed.

Reviewer 5: How is it dedicated to ESM? `--type=commonjs` says to load a file as CommonJS . . . ?

Reviewer 4: i believe joyee meant the esm loader, not source text modules.

Author: I'm going through and updating to `--entry-type` for now, and will punt back to the modules team regarding the change.

@joyeecheung where this effects both loaders is the behavior of `node entry.js`.

While it is very likely the default for node is never going to change folks could guard for the future with `node --entry-type=commonjs entry.js` to ensure they will always get the CJS loader.

> ### `--entry-type=type`

Author: fixed in a2f584a40b

Track heap object allocations for heap snapshots.

> ### `-m`, `--type=type`

Reviewer 6: ```suggestion
### `--type=type`
```

Used with `--experimental-modules`, this configures Node.js to interpret the
initial entry point as CommonJS or as an ES module.

Valid values are `"commonjs"` and `"module"`. The default is to infer from
the file extension and the `"type"` field in the nearest parent `package.json`.

Works for executing a file as well as `--eval`, `--print`, `STDIN`.

> `-m` is an alias for `--type=module`.

Reviewer 6: ```suggestion
```

[debugger]: debugger.html
[debugging security implications]: https://nodejs.org/en/docs/guides/debugging-getting-started/#security-implications
[emit_warning]: process.html#process_process_emitwarning_warning_type_code_ctor
[experimental ECMAScript Module]: esm.html#esm_resolve_hook

> [customizing esm specifier resolution]: esm.html#esm_customizing_esm_specifier_resolution_algorithm

Author: fixed in 858060e0a6
